src/box/iproto.cc

```diff
@@ -2483,6 +2516,156 @@ tx_process_replication(struct cmsg *m)
+		 * replica id. Ignore the header received over network.
+		 */
+		msg->dml.header = NULL;
+		assert(type < sizeof(iproto_thread->dml_route) /
+			      sizeof(*iproto_thread->dml_route));
+		iproto_thread->dml_route[type]->f(m);
+		break;
+	case IPROTO_BEGIN:
+		if (xrow_decode_begin(&msg->header, &msg->begin) != 0)
+			return -1;
+		iproto_thread->begin_route->f(m);
+		break;
+	case IPROTO_COMMIT:
+		iproto_thread->commit_route->f(m);
+		break;
+	case IPROTO_ROLLBACK:
+		iproto_thread->rollback_route->f(m);
+		break;
+	case IPROTO_CALL_16:
+	case IPROTO_CALL:
+	case IPROTO_EVAL:
+		if (xrow_decode_call(&msg->header, &msg->call))
+			return -1;
+		iproto_thread->call_route->f(m);
+		break;
+	case IPROTO_WATCH:
+	case IPROTO_UNWATCH:
+		if (xrow_decode_watch(&msg->header, &msg->watch) != 0)
+			return -1;
+		iproto_thread->misc_route->f(m);
+		break;
+	case IPROTO_EXECUTE:
+	case IPROTO_PREPARE:
+		if (xrow_decode_sql(&msg->header, &msg->sql) != 0)
+			return -1;
+		iproto_thread->sql_route->f(m);
+		break;
+	case IPROTO_PING:
+		iproto_thread->misc_route->f(m);
+		break;
+	case IPROTO_ID:
+		if (xrow_decode_id(&msg->header, &msg->id) != 0)
+			return -1;
+		iproto_thread->misc_route->f(m);
+		break;
+	case IPROTO_VOTE_DEPRECATED:
+	case IPROTO_VOTE:
+		iproto_thread->misc_route->f(m);
+		break;
+	case IPROTO_AUTH:
+		if (xrow_decode_auth(&msg->header, &msg->auth))
+			return -1;
+		iproto_thread->misc_route->f(m);
+		break;
+	default:
+		diag_set(ClientError, ER_UNKNOWN_REQUEST_TYPE, type);
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Process a request using the overridden handler (or the unknown request
+ * handler as a last resort). We assume that the IPROTO thread checked the
+ * handler availability.
+ */
+static void
+tx_process_override(struct cmsg *m)
+{
+	struct iproto_msg *msg = tx_accept_msg(m);
+	struct obuf *out;
+	struct obuf_svp svp;
+	iproto_handler *node;
+	const char *header;
+	const char *header_end;
+	const char *body;
+	const char *body_end;
+	size_t body_len;
+
+	mh_int_t k = mh_rhvt_find(tx_rhvt, msg->header.type, NULL);
+	if (k == mh_end(tx_rhvt)) {
+		k = mh_rhvt_find(tx_rhvt, IPROTO_UNKNOWN, NULL);
+		assert(k != mh_end(tx_rhvt));
+	}
+	node = mh_rhvt_node(tx_rhvt, k);
+	header = msg->reqstart;
+	mp_decode_uint(&header);
+
+	const char *empty_body = "\x80";
+	if (msg->header.bodycnt == 0) {
+		body = empty_body;
+		body_end = empty_body + 1;
+		body_len = 0;
+	} else {
+		assert(msg->header.bodycnt == 1);
+		body_len = msg->header.body[0].iov_len;
+		body = (const char *)msg->header.body[0].iov_base;
+		body_end = body + msg->header.body[0].iov_len;
+	}
+	header_end = msg->reqstart + msg->len - body_len;
+
+	switch (node->callback(header, header_end, body, body_end, node->ctx)) {
+	case IPROTO_HANDLER_OK: {
+		iproto_wpos_create(&msg->wpos, msg->connection->tx.p_obuf);
+		struct obuf_svp empty =
+			obuf_create_svp(msg->connection->tx.p_obuf);
+		tx_end_msg(msg, &empty);
```

Comment:
> OTOH flightrec shouldn't log empty requests.

But, apparently, it does when replication requests are processed: https://github.com/tarantool/tarantool/blob/706bfea476a013bb8cf7566666699eb331876d7f/src/box/iproto.cc#L2482-L2483

> Instead, it should log whatever box_iproto_send sends.

Fixed. Should this be tested?

Reply:
> Should this be tested?

Ideally, yes, but this can only be done in EE. Please open a ticket in the EE repository to test how flightrec works with IPROTO override.